whylog/constraints/__init__.py

```diff
@@ -1,43 +1,200 @@
+        if (self.MIN_GROUPS_COUNT is not None and groups_count < self.MIN_GROUPS_COUNT)\
+                or (self.MAX_GROUPS_COUNT is not None and groups_count > self.MAX_GROUPS_COUNT):
+            raise ConstructorGroupsCountError(
+                self.TYPE, len(self.groups), self.MIN_GROUPS_COUNT, self.MAX_GROUPS_COUNT
+            )
 
-class TimeConstraint(AbstractConstraint):
-    def __init__(self, line_earlier, line_later, min_delta=None, max_delta=None):
+    @abstractmethod
+    def _check_constructor_params(self):
         pass
 
+    def convert_to_user_constraint_intent(self):
+        """
+        Converts constraint to UserConstraintIntent object.
 
-class IdenticalIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+        For Teacher and Config use while saving constraint into Whylog knowledge base.
+        """
+        return UserConstraintIntent(self.TYPE, self.groups, self.params)
 
+    @classmethod
+    def get_param_names(cls):
+        """
+        Returns names of constraint additional params.
 
-class AnyValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+        For Front to display param names to user and then ask user for param contents.
+        """
+        return cls.PARAMS
 
+    @classmethod
+    def get_groups_count(cls):
+        """
+        Returns minimal and maximal count of groups needed to create constraint.
 
-class DifferentValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+        For Front to ask user for proper count of groups.
+        2, None - at least 2 groups
+        2, 2 - exactly 2 groups
+        """
+        return cls.MIN_GROUPS_COUNT, cls.MAX_GROUPS_COUNT
 
+    def verify(cls, group_contents, param_dict):
+        """
+        Verifies constraint for given params in param_dict and groups contents.
+
+        :param param_dict: dict of additional params of constraint
+        :param groups: list of groups contents,
+
+        For LogReader and Teacher verification.
+        It must be optimized as well as possible (for LogReader).
+        """
 
-class ConstIntervals(AbstractConstraint):
-    def __init__(self, intervals):
         pass
 
 
-class ValueDeltaIntervals(AbstractConstraint):
-    def __init__(self, interval_lower, interval_greater, min_delta=None, max_delta=None):
+class TimeConstraint(AbstractConstraint):
+    """
+    Time delta between two dates must be greater than 'max_delta' and lower than 'min_delta'
+    """
+
+    TYPE = ConstraintType.TIME_DELTA
+
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = 2
+
+    MIN_DELTA = 'min_delta'
+    MAX_DELTA = 'max_delta'
+
+    PARAMS = sorted([MIN_DELTA, MAX_DELTA])
+
+    def __init__(self, groups, param_dict):
         """
-        Sets Minimum and maximum difference between values of params (if values are numbers).
+        :param groups: First element of groups is a group with earlier date.
```

Comment:
I can't throw exception if first group is not earlier, because while constructing I don't have groups contents. Teacher will run verify method on users constraint to check such issues.